### PR TITLE
Expose "key address" to normalization callbacks

### PR DESCRIPTION
**Pull request by happz**
Keys belong to various objects like step phases or fmf ids, and tmt can do better job when reporting unexpected input values. Another step is to expose the location of the key to normalization callbacks and a dedicated exception for reporting the violations in a unified manner.

The key address is already used by schema validation when reporting validation errors, and while it's often subpar, it enhances error messages a bit.

In the follow-up patches, improving the addresses will happen - for example, sometimes a class name is printed instead of a fmf node name, usually because the node name is not available.

### CODE REVIEW

**Comment by idorax** [comment]
Hi @happz, do you plan get this PR merged soon? PR https://github.com/teemtee/tmt/pull/1905 of mine has a dependency on this one, I'll update it once your PR is merged, thanks!

**Comment by idorax** [comment]
Hi @happz, your PR looks good to me.

And I also tested your patch by updating `tmt/base.py` because of PR https://github.com/teemtee/tmt/pull/1905, the test result looks good. 

* Patch
```patch
(tmt) huanli@kvm-01-guest01:dev$ git branch
* dev.huanli.1881.20230417
  main
  normalize-include-key-address
(tmt) huanli@kvm-01-guest01:dev$ git diff
diff --git a/tmt/base.py b/tmt/base.py
index 7347dc4..7f316a3 100644
--- a/tmt/base.py
+++ b/tmt/base.py
@@ -766,7 +766,11 @@ class Test(Core, tmt.export.Exportable['Test']):
         # TODO: mandatory schema validation would remove the need for Optional...
         # `test` is mandatory, must exist, so how to initialize if it's missing :(
         if value is None:
-            return ShellScript('')
+            raise tmt.utils.NormalizationError(key_address, value, 'string')
+
+        # Raise an error if value is not a string, e.g. False or True
+        if not isinstance(value, str):
+            raise tmt.utils.NormalizationError(key_address, value, 'string')
 
         return ShellScript(value)
```

* Test result
```bash
(tmt) huanli@kvm-01-guest01:foo$ pwd 
/home/huanli/dev/1001869/foo/tests/foo
(tmt) huanli@kvm-01-guest01:foo$ cat main.fmf 
summary: Concise summary describing what the test does
test: false 
framework: beakerlib
(tmt) huanli@kvm-01-guest01:foo$ tmt run -avvv provision -h local
/var/tmp/tmt/run-001
Found 1 plan.

/plans
summary: Basic smoke test
    discover
        how: fmf
        order: 50
        directory: /home/huanli/dev/1001869/foo
                warn: /tests/foo:test - False is not of type 'string'
    finish
        prune: Prune plan workdir '/var/tmp/tmt/run-001/plans'.
        summary: 0 tasks completed

Field "/tests/foo:test" can be string, "bool" found.
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  Looks good !!!
```

File: tests/unit/test_base.py [hunk 1]
```diff
@@ -139,7 +139,10 @@ def test_link():
             note=fmf_id['note'])]
 
     # Invalid links and relations
-    with pytest.raises(SpecificationError, match='Invalid link'):
+    with pytest.raises(
+            SpecificationError,
+            match="Field 'link' can be string, fmf id or list of their commbinations,"
```
Comment (lukaszachy) [inline tests/unit/test_base.py:144]: Typo: double 'm'

File: tmt/base.py [hunk 2]
```diff
@@ -3514,10 +3535,9 @@ def __init__(self, *, data: Optional[_RawLinks] = None):
 
         # TODO: this should not happen with mandatory validation
         if data is not None and not isinstance(data, (str, dict, list)):
-            raise tmt.utils.SpecificationError(
-                "Invalid link specification "
-                "(should be a string, fmf id or list of their combinations), "
-                f"got '{type(data).__name__}'.")
+            # TODO: deliver better key address, needs to know the parent
+            raise tmt.utils.NormalizationError(
+                'link', data, 'string, fmf id or list of their commbinations')
```
Comment (lukaszachy) [inline tmt/base.py:3540]: Typo: double 'm'

**Review by lukaszachy — COMMENTED** [review]
Just typos found.

File: tests/unit/test_base.py (hunk 1 above)
Comment (happz) [reply tests/unit/test_base.py]: Fixed in 9529a776134c84fe8c7428edbfe79f2eb70cfaa4

File: tmt/base.py (hunk 2 above)
Comment (happz) [reply tmt/base.py]: Fixed in 9529a776134c84fe8c7428edbfe79f2eb70cfaa4

**Comment by lukaszachy** [comment]
Fix for testing-farm:fedora-38-x86_64:internal-plugins  in progress

**Review by lukaszachy — APPROVED** [review]

File: tests/core/fmf-id/test.sh [hunk 3]
```diff
@@ -14,7 +14,7 @@ rlJournalStart
 
         rlRun -s "tmt plan -vvvv show /plan-with-invalid-ref" 2
         rlAssertGrep "warn: /plan-with-invalid-ref:discover .* is not valid under any of the given schemas" $rlRun_LOG
-        rlAssertGrep "Failed to load step data for DiscoverFmfStepData: The 'ref' field must be a string, got 'int'." $rlRun_LOG
+        rlAssertGrep "Failed to load step data for DiscoverFmfStepData: Field 'DiscoverFmfStepData:ref' can be string, 'int' found." $rlRun_LOG
```
Comment (thrix) [inline tests/core/fmf-id/test.sh:17]: personally, must feels more correct? can it be also something else?

File: tests/core/fmf-id/test.sh [hunk 4]
```diff
@@ -28,7 +28,8 @@ rlJournalStart
         rlAssertGrep "some-package" $rlRun_LOG
 
         rlRun -s "tmt test -vvvv show /test-with-invalid-ref" 2
-        rlAssertGrep "The 'ref' field must be a string, got 'int'." $rlRun_LOG
+        rlAssertGrep "warn: /test-with-invalid-ref:require .* is not valid under any of the given schemas" $rlRun_LOG
+        rlAssertGrep "Field 'ref' can be not set or string, 'int' found." $rlRun_LOG
```
Comment (thrix) [inline tests/core/fmf-id/test.sh:32]: `can be not set or string`? well, I would rather go with

```
Field 'ref' must be unset or a string, 'int' found.
```

Would `must` be more correct here? Can  it be something else?

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/base.py [hunk 5]
```diff
@@ -161,8 +161,8 @@ def from_spec(cls, raw: _RawFmfId) -> 'FmfId':
         # TODO: with mandatory validation, this can go away.
         ref = raw.get('ref', None)
         if not isinstance(ref, (type(None), str)):
-            raise tmt.utils.SpecificationError(
-                f"The 'ref' field must be a string, got '{type(ref).__name__}'.")
+            # TODO: deliver better key address
+            raise tmt.utils.NormalizationError('ref', ref, 'not set or string')
```
Comment (janhavlin) [inline tmt/base.py:165]: This would result in a message:
```
Field 'ref' can be not set or string, 'int' found.
```
I think it is gramatically incorrect and ambiguous, it could be interpreted as `it can be not (set or string)` which is unintended.

I'd say `unset` is better than `not set`:
```
Field 'ref' can be unset or string, 'int' found.
```

File: tmt/utils.py [hunk 6]
```diff
@@ -1281,6 +1281,39 @@ def __init__(
...
+
+        :param key_address: the key in question, preferrably with detailed location,
+            e.g. ``/plans/foo:discover[0].tests``.
+        :param raw_value: input value, the one that failed the normalization.
+        :param expected_type: string description of expected, allowed types, as
+            a hint in the error message.
+        """
+
+        super().__init__(
+            f"Field '{key_address}' can be {expected_type}, '{type(raw_value).__name__}' found.",
```
Comment (janhavlin) [inline tmt/utils.py:1308]: `must` might be a more suitable word rather than `can`?

File: tests/core/fmf-id/test.sh (hunk 3 above)
Comment (happz) [reply tests/core/fmf-id/test.sh]: I agree, changed in ede71004.

File: tests/core/fmf-id/test.sh (hunk 4 above)
Comment (happz) [reply tests/core/fmf-id/test.sh]: Fixed in ede71004.

File: tmt/base.py (hunk 5 above)
Comment (happz) [reply tmt/base.py]: Fixed in ede71004.

**Review by thrix — APPROVED** [review]
LGTM, ty!

File: tmt/steps/prepare/install.py [hunk 7]
```diff
@@ -522,7 +521,7 @@ class PrepareInstallData(tmt.steps.prepare.PrepareStepData):
         )
 
 
-@tmt.steps.provides_method('install')
+@ tmt.steps.provides_method('install')
```
Comment (psss) [inline tmt/steps/prepare/install.py:524]: Was this change intentional?

Comment (happz) [reply tmt/steps/prepare/install.py:524]: Oops, nope, that does not look like an intention. It shouldn't have any effect, but it's not supposed to happen.

Comment (psss) [reply tmt/steps/prepare/install.py:524]: Ok, we can fix with some other patch.
